src/main/java/com/gregtechceu/gtceu/api/machine/trait/NotifiableFluidTank.java

```diff
@@ -116,42 +117,71 @@ public List<FluidIngredient> handleRecipeInner(IO io, GTRecipe recipe, List<Flui
+                        amount = provider.getCountProvider().getMaxValue();
+                    } else {
+                        fluids = provider.getStacks();
+                        if (fluids.length == 0 || fluids[0].isEmpty()) {
+                            it.remove();
+                            continue;
+                        }
+                        amount = fluids[0].getAmount();
+                    }
+                } else {
+                    fluids = ingredient.getStacks();
+                    if (fluids.length == 0 || fluids[0].isEmpty()) {
+                        it.remove();
+                        continue;
                     }
+                    amount = ingredient.getAmount();
                 }
-                if (existing != null) {
-                    FluidStack output = fluids[0].copy();
-                    output.setAmount(amount);
-                    int filled = existing.fill(output, action);
-                    if (filled > 0) {
-                        visited[tank] = output.copy();
-                        // shortcut for oldAmount + filled (wow what an idea)
-                        visited[tank].setAmount(existing.getFluidAmount());
-                        changed = true;
+
+                if (!allowSameFluids) {
+                    CustomFluidTank existing = null;
+                    int tank = 0;
+                    for (int i = 0; i < storages.length; ++i) {
+                        var storage = storages[i];
+                        if (!storage.getFluid().isEmpty() && storage.getFluid().isFluidEqual(fluids[0])) {
+                            existing = storage;
+                            tank = i;
+                            break;
+                        }
                     }
-                    amount -= filled;
+                    if (existing != null) {
+                        FluidStack output = fluids[0].copy();
+                        output.setAmount(amount);
+                        int filled = existing.fill(output, action);
+                        if (filled > 0) {
+                            visited[tank] = output.copy();
+                            // shortcut for oldAmount + filled (wow what an idea)
+                            visited[tank].setAmount(existing.getFluidAmount());
+                            changed = true;
+                        }
+                        amount -= filled;
 
-                    if (amount > 0) ingredient.setAmount(amount);
-                    else it.remove();
-                    // Continue to next ingredient regardless of if we filled this ingredient completely
+                        if (amount > 0) ingredient.setAmount(amount);
+                        else it.remove();
+                        // Continue to next ingredient regardless of if we filled this ingredient completely
+                        continue;
+                    }
+                }
+            } else {
```

Comment:
instead of duplicating this `else ...` on line 140, do `FluidStack[] fluids = null;` on line 120 and check if it's null here (and remove the `else`, obviously)
so something like:
```suggestion
            }
            if (fluids == null) {
```